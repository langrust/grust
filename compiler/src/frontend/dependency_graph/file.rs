use petgraph::algo::toposort;
use petgraph::graphmap::DiGraphMap;

prelude! {
    error::{Error, TerminationError},
    hir::file::File,
    symbol_table::SymbolTable,

}

impl File {
    /// Generate dependency graph for every nodes/component.
    pub fn generate_dependency_graphs(
        &mut self,
        symbol_table: &SymbolTable,
        errors: &mut Vec<Error>,
    ) -> Result<(), TerminationError> {
        let File { nodes, .. } = self;

        // initialize dictionariy for reduced graphs
        let mut nodes_reduced_graphs = HashMap::new();

        // create graph of nodes
        let mut nodes_graph = DiGraphMap::new();
        nodes
            .iter()
            .for_each(|node| node.add_node_dependencies(&mut nodes_graph));

        // sort nodes according to their dependencies
        let sorted_nodes = toposort(&nodes_graph, None).map_err(|node| {
            let error = Error::NotCausalNode {
                node: symbol_table.get_name(node.node_id()).clone(),
                location: self.location.clone(),
            };
            errors.push(error);
            TerminationError
        })?;
        nodes.sort_by(|n1, n2| {
            let index1 = sorted_nodes
                .iter()
                .position(|id| *id == n1.id)
                .expect("should be in sorted list");
            let index2 = sorted_nodes
                .iter()
                .position(|id| *id == n2.id)
                .expect("should be in sorted list");

            Ord::cmp(&index2, &index1)
        });

        // ordered nodes complete their dependency graphs
        nodes
            .iter_mut()
            .map(|node| node.compute_dependencies(symbol_table, &mut nodes_reduced_graphs, errors))
            .collect::<Vec<Result<(), TerminationError>>>()
            .into_iter()
            .collect::<Result<(), TerminationError>>()?;

        Ok(())
    }
}
