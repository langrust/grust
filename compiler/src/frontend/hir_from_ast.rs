prelude! {}

/// HIR Component construction from AST Component
pub mod component;
/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
pub mod typ;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST<Ctxt> {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(self, ctxt: &mut Ctxt) -> TRes<Self::HIR>;
}

pub struct SimpleCtxt<'a> {
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct LocCtxt<'a> {
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct PatLocCtxt<'a> {
    pub pat: Option<&'a ast::stmt::Pattern>,
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
impl<'a> SimpleCtxt<'a> {
    pub fn new(syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { syms, errors }
    }
    pub fn add_loc<'b>(&'b mut self, loc: &'b Location) -> LocCtxt<'b> {
        LocCtxt::new(loc, self.syms, self.errors)
    }
    pub fn add_pat_loc<'b>(
        &'b mut self,
        pat: Option<&'b ast::stmt::Pattern>,
        loc: &'b Location,
    ) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, loc, self.syms, self.errors)
    }
}
impl<'a> LocCtxt<'a> {
    pub fn new(loc: &'a Location, syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { loc, syms, errors }
    }
    pub fn add_pat<'b>(&'b mut self, pat: Option<&'b ast::stmt::Pattern>) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, self.loc, self.syms, self.errors)
    }
}
impl<'a> PatLocCtxt<'a> {
    pub fn new(
        pat: Option<&'a ast::stmt::Pattern>,
        loc: &'a Location,
        syms: &'a mut SymbolTable,
        errors: &'a mut Vec<Error>,
    ) -> Self {
        Self {
            pat,
            loc,
            syms,
            errors,
        }
    }
    pub fn remove_pat<'b>(&'b mut self) -> LocCtxt<'b> {
        LocCtxt::new(self.loc, self.syms, self.errors)
    }
    pub fn set_pat(
        &mut self,
        pat: Option<&'a ast::stmt::Pattern>,
    ) -> Option<&'a ast::stmt::Pattern> {
        std::mem::replace(&mut self.pat, pat)
    }
}
