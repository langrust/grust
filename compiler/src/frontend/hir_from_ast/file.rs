prelude! {
    ast::Ast,
    hir::interface::Interface,
}

use super::HIRFromAST;

impl HIRFromAST for Ast {
    type HIR = hir::File;

    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        // initialize symbol table with builtin operators
        symbol_table.initialize();

        // store elements in symbol table
        self.store(symbol_table, errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Function(function) => {
                        functions.push(function.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Typedef(typedef) => {
                        typedefs.push(typedef.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Service(service) => {
                        services.push(service.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Import(import) => imports.push(
                        import
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
