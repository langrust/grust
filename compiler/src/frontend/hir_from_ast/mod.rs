prelude! {}

/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR>;
}

impl HIRFromAST for ast::Component {
    type HIR = hir::Node;

    // precondition: node and its signals are already stored in symbol table
    // postcondition: construct HIR node and check identifiers good use
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        let ast::Component {
            ident,
            contract,
            equations,
            ..
        } = self;
        let name = ident.to_string();
        let location = Location::default();
        let id = symbol_table.get_node_id(&name, false, location.clone(), errors)?;

        // create local context with all signals
        symbol_table.local();
        symbol_table.restore_context(id);
        symbol_table.enter_in_node(id);

        let statements = equations
            .into_iter()
            .map(|equation| equation.hir_from_ast(symbol_table, errors))
            .collect::<TRes<Vec<_>>>()?;
        let contract = contract.hir_from_ast(symbol_table, errors)?;

        symbol_table.leave_node();
        symbol_table.global();

        Ok(hir::Node {
            id,
            statements,
            contract,
            location,
            graph: graph::DiGraphMap::new(),
            memory: hir::Memory::new(),
        })
    }
}
