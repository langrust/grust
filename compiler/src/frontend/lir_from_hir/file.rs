use crate::{
    hir::file::File,
    lir::{item::Item, project::Project},
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for File {
    type LIR = Project;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let File {
            typedefs,
            functions,
            nodes,
            interface,
            ..
        } = self;

        let mut typedefs = typedefs
            .into_iter()
            .map(|typedef| typedef.lir_from_hir(symbol_table))
            .collect();
        let mut functions = functions
            .into_iter()
            .map(|function| function.lir_from_hir(symbol_table))
            .map(Item::Function)
            .collect();
        let mut state_machines = nodes
            .into_iter()
            .map(|node| node.lir_from_hir(symbol_table))
            .map(Item::StateMachine)
            .collect();
        let execution_machine = interface.lir_from_hir(symbol_table);

        let mut items = vec![];
        items.append(&mut typedefs);
        items.append(&mut functions);
        items.append(&mut state_machines);
        items.push(Item::ExecutionMachine(execution_machine));

        Project { items }
    }
}
