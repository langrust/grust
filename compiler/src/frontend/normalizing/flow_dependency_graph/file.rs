prelude! { just
    hir::File,
}

impl File {
    /// Generate dependency graphs for the interface.
    #[inline]
    pub fn generate_flows_dependency_graphs(&mut self) {
        self.interface.generate_flows_dependency_graphs()
    }
}
