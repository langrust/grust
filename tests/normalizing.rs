use codespan_reporting::files::{Files, SimpleFiles};

use grustine::ast::file::File;
use grustine::frontend::hir_from_ast::file::hir_from_ast;
use grustine::parser::langrust;

#[test]
fn normalize_counter() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/counter.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(counter_id, &files.source(counter_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_blinking() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/blinking.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_button_management() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/button_management.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_button_management_condition_match() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/button_management_condition_match.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}

#[test]
fn normalize_button_management_using_function() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/button_management_using_function.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();

    file.normalize(&mut errors).unwrap();
    insta::assert_yaml_snapshot!(file);
}
