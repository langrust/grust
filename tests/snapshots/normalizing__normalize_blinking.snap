---
source: tests/normalizing.rs
expression: file
---
typedefs: []
functions: []
nodes:
  - id: counter
    is_component: false
    inputs:
      - - res
        - Boolean
      - - tick
        - Boolean
    unscheduled_equations:
      inc:
        scope: Local
        id: inc
        signal_type: Integer
        expression:
          MapApplication:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Integer
                      - Integer
                    - Integer
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: tick
                    scope: Input
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - tick
                      - 0
              - Constant:
                  constant:
                    Integer: 1
                  typing: Integer
                  location: ~
                  dependencies: []
              - Constant:
                  constant:
                    Integer: 0
                  typing: Integer
                  location: ~
                  dependencies: []
            typing: Integer
            location: ~
            dependencies:
              - - tick
                - 0
        location: ~
      o:
        scope: Output
        id: o
        signal_type: Integer
        expression:
          MapApplication:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Integer
                      - Integer
                    - Integer
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Input
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - res
                      - 0
              - Constant:
                  constant:
                    Integer: 0
                  typing: Integer
                  location: ~
                  dependencies: []
              - MapApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Integer
                            - Integer
                          - Integer
                      location: ~
                  inputs:
                    - MapApplication:
                        function_expression:
                          Call:
                            id: (_)
                            typing:
                              Abstract:
                                - - Integer
                                - Integer
                            location: ~
                        inputs:
                          - FollowedBy:
                              constant:
                                Integer: 0
                              expression:
                                SignalCall:
                                  signal:
                                    id: o
                                    scope: Output
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - o
                                      - 0
                              typing: Integer
                              location: ~
                              dependencies:
                                - - o
                                  - 1
                        typing: Integer
                        location: ~
                        dependencies:
                          - - o
                            - 1
                    - SignalCall:
                        signal:
                          id: inc
                          scope: Local
                        typing: Integer
                        location: ~
                        dependencies:
                          - - inc
                            - 0
                  typing: Integer
                  location: ~
                  dependencies:
                    - - o
                      - 1
                    - - inc
                      - 0
            typing: Integer
            location: ~
            dependencies:
              - - res
                - 0
              - - o
                - 1
              - - inc
                - 0
        location: ~
    unitary_nodes:
      o:
        node_id: counter
        output_id: o
        inputs:
          - - res
            - Boolean
          - - tick
            - Boolean
        equations:
          - scope: Local
            id: inc
            signal_type: Integer
            expression:
              MapApplication:
                function_expression:
                  Call:
                    id: if_then_else
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                inputs:
                  - SignalCall:
                      signal:
                        id: tick
                        scope: Input
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - tick
                          - 0
                  - Constant:
                      constant:
                        Integer: 1
                      typing: Integer
                      location: ~
                      dependencies: []
                  - Constant:
                      constant:
                        Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
                typing: Integer
                location: ~
                dependencies:
                  - - tick
                    - 0
            location: ~
          - scope: Output
            id: o
            signal_type: Integer
            expression:
              MapApplication:
                function_expression:
                  Call:
                    id: if_then_else
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                inputs:
                  - SignalCall:
                      signal:
                        id: res
                        scope: Input
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - res
                          - 0
                  - Constant:
                      constant:
                        Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
                  - MapApplication:
                      function_expression:
                        Call:
                          id: " + "
                          typing:
                            Abstract:
                              - - Integer
                                - Integer
                              - Integer
                          location: ~
                      inputs:
                        - MapApplication:
                            function_expression:
                              Call:
                                id: (_)
                                typing:
                                  Abstract:
                                    - - Integer
                                    - Integer
                                location: ~
                            inputs:
                              - SignalCall:
                                  signal:
                                    id: memo
                                    scope: Memory
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - memo
                                      - 0
                            typing: Integer
                            location: ~
                            dependencies:
                              - - memo
                                - 0
                        - SignalCall:
                            signal:
                              id: inc
                              scope: Local
                            typing: Integer
                            location: ~
                            dependencies:
                              - - inc
                                - 0
                      typing: Integer
                      location: ~
                      dependencies:
                        - - memo
                          - 0
                        - - inc
                          - 0
                typing: Integer
                location: ~
                dependencies:
                  - - res
                    - 0
                  - - memo
                    - 0
                  - - inc
                    - 0
            location: ~
        memory:
          buffers:
            memo:
              typing: Integer
              initial_value:
                Integer: 0
              expression:
                SignalCall:
                  signal:
                    id: o
                    scope: Output
                  typing: Integer
                  location: ~
                  dependencies:
                    - - o
                      - 0
          called_nodes: {}
        location: ~
        graph:
          vertices:
            inc:
              id: inc
              value: White
              neighbors:
                - id: tick
                  weight: 0
            memo:
              id: memo
              value: White
              neighbors: []
            o:
              id: o
              value: White
              neighbors:
                - id: res
                  weight: 0
                - id: memo
                  weight: 0
                - id: inc
                  weight: 0
            res:
              id: res
              value: White
              neighbors: []
            tick:
              id: tick
              value: White
              neighbors: []
    location: ~
    graph:
      vertices:
        inc:
          id: inc
          value: Black
          neighbors:
            - id: tick
              weight: 0
        o:
          id: o
          value: Black
          neighbors:
            - id: res
              weight: 0
            - id: o
              weight: 1
            - id: inc
              weight: 0
        res:
          id: res
          value: Black
          neighbors: []
        tick:
          id: tick
          value: Black
          neighbors: []
  - id: blinking
    is_component: false
    inputs:
      - - tick_number
        - Integer
    unscheduled_equations:
      counter:
        scope: Local
        id: counter
        signal_type: Integer
        expression:
          NodeApplication:
            node: counter
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Local
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - res
                      - 0
              - Constant:
                  constant:
                    Boolean: true
                  typing: Boolean
                  location: ~
                  dependencies: []
            signal: o
            typing: Integer
            location: ~
            dependencies:
              - - res
                - 0
              - - res
                - 1
        location: ~
      on_off:
        scope: Local
        id: on_off
        signal_type: Boolean
        expression:
          MapApplication:
            function_expression:
              TypedAbstraction:
                inputs:
                  - - t
                    - Boolean
                  - - b
                    - Boolean
                expression:
                  Application:
                    function_expression:
                      Call:
                        id: if_then_else
                        typing:
                          Abstract:
                            - - Boolean
                              - Boolean
                              - Boolean
                            - Boolean
                        location: ~
                    inputs:
                      - Call:
                          id: t
                          typing: Boolean
                          location: ~
                      - Application:
                          function_expression:
                            Call:
                              id: "!"
                              typing:
                                Abstract:
                                  - - Boolean
                                  - Boolean
                              location: ~
                          inputs:
                            - Call:
                                id: b
                                typing: Boolean
                                location: ~
                          typing: Boolean
                          location: ~
                      - Call:
                          id: b
                          typing: Boolean
                          location: ~
                    typing: Boolean
                    location: ~
                typing:
                  Abstract:
                    - - Boolean
                      - Boolean
                    - Boolean
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Local
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - res
                      - 0
              - FollowedBy:
                  constant:
                    Boolean: true
                  expression:
                    SignalCall:
                      signal:
                        id: on_off
                        scope: Local
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - on_off
                          - 0
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - on_off
                      - 1
            typing: Boolean
            location: ~
            dependencies:
              - - res
                - 0
              - - on_off
                - 1
        location: ~
      res:
        scope: Local
        id: res
        signal_type: Boolean
        expression:
          FollowedBy:
            constant:
              Boolean: true
            expression:
              MapApplication:
                function_expression:
                  Call:
                    id: (_)
                    typing:
                      Abstract:
                        - - Boolean
                        - Boolean
                    location: ~
                inputs:
                  - MapApplication:
                      function_expression:
                        Call:
                          id: " == "
                          typing:
                            Abstract:
                              - - Integer
                                - Integer
                              - Boolean
                          location: ~
                      inputs:
                        - MapApplication:
                            function_expression:
                              Call:
                                id: " + "
                                typing:
                                  Abstract:
                                    - - Integer
                                      - Integer
                                    - Integer
                                location: ~
                            inputs:
                              - SignalCall:
                                  signal:
                                    id: counter
                                    scope: Local
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - counter
                                      - 0
                              - Constant:
                                  constant:
                                    Integer: 1
                                  typing: Integer
                                  location: ~
                                  dependencies: []
                            typing: Integer
                            location: ~
                            dependencies:
                              - - counter
                                - 0
                        - SignalCall:
                            signal:
                              id: tick_number
                              scope: Input
                            typing: Integer
                            location: ~
                            dependencies:
                              - - tick_number
                                - 0
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - counter
                          - 0
                        - - tick_number
                          - 0
                typing: Boolean
                location: ~
                dependencies:
                  - - counter
                    - 0
                  - - tick_number
                    - 0
            typing: Boolean
            location: ~
            dependencies:
              - - counter
                - 1
              - - tick_number
                - 1
        location: ~
      status:
        scope: Output
        id: status
        signal_type: Integer
        expression:
          MapApplication:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Integer
                      - Integer
                    - Integer
                location: ~
            inputs:
              - SignalCall:
                  signal:
                    id: on_off
                    scope: Local
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - on_off
                      - 0
              - MapApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Integer
                            - Integer
                          - Integer
                      location: ~
                  inputs:
                    - SignalCall:
                        signal:
                          id: counter
                          scope: Local
                        typing: Integer
                        location: ~
                        dependencies:
                          - - counter
                            - 0
                    - Constant:
                        constant:
                          Integer: 1
                        typing: Integer
                        location: ~
                        dependencies: []
                  typing: Integer
                  location: ~
                  dependencies:
                    - - counter
                      - 0
              - Constant:
                  constant:
                    Integer: 0
                  typing: Integer
                  location: ~
                  dependencies: []
            typing: Integer
            location: ~
            dependencies:
              - - on_off
                - 0
              - - counter
                - 0
        location: ~
    unitary_nodes:
      status:
        node_id: blinking
        output_id: status
        inputs:
          - - tick_number
            - Integer
        equations:
          - scope: Local
            id: res
            signal_type: Boolean
            expression:
              SignalCall:
                signal:
                  id: memres
                  scope: Memory
                typing: Boolean
                location: ~
                dependencies:
                  - - memres
                    - 0
            location: ~
          - scope: Local
            id: x
            signal_type: Boolean
            expression:
              Constant:
                constant:
                  Boolean: true
                typing: Boolean
                location: ~
                dependencies: []
            location: ~
          - scope: Local
            id: counter
            signal_type: Integer
            expression:
              UnitaryNodeApplication:
                id: counterocounter
                node: counter
                signal: o
                inputs:
                  - - res
                    - SignalCall:
                        signal:
                          id: res
                          scope: Local
                        typing: Boolean
                        location: ~
                        dependencies:
                          - - res
                            - 0
                  - - tick
                    - SignalCall:
                        signal:
                          id: x
                          scope: Local
                        typing: Boolean
                        location: ~
                        dependencies:
                          - - x
                            - 0
                typing: Integer
                location: ~
                dependencies:
                  - - res
                    - 0
                  - - x
                    - 0
            location: ~
          - scope: Local
            id: on_off
            signal_type: Boolean
            expression:
              MapApplication:
                function_expression:
                  TypedAbstraction:
                    inputs:
                      - - t
                        - Boolean
                      - - b
                        - Boolean
                    expression:
                      Application:
                        function_expression:
                          Call:
                            id: if_then_else
                            typing:
                              Abstract:
                                - - Boolean
                                  - Boolean
                                  - Boolean
                                - Boolean
                            location: ~
                        inputs:
                          - Call:
                              id: t
                              typing: Boolean
                              location: ~
                          - Application:
                              function_expression:
                                Call:
                                  id: "!"
                                  typing:
                                    Abstract:
                                      - - Boolean
                                      - Boolean
                                  location: ~
                              inputs:
                                - Call:
                                    id: b
                                    typing: Boolean
                                    location: ~
                              typing: Boolean
                              location: ~
                          - Call:
                              id: b
                              typing: Boolean
                              location: ~
                        typing: Boolean
                        location: ~
                    typing:
                      Abstract:
                        - - Boolean
                          - Boolean
                        - Boolean
                    location: ~
                inputs:
                  - SignalCall:
                      signal:
                        id: res
                        scope: Local
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - res
                          - 0
                  - SignalCall:
                      signal:
                        id: memon_off
                        scope: Memory
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - memon_off
                          - 0
                typing: Boolean
                location: ~
                dependencies:
                  - - res
                    - 0
                  - - memon_off
                    - 0
            location: ~
          - scope: Output
            id: status
            signal_type: Integer
            expression:
              MapApplication:
                function_expression:
                  Call:
                    id: if_then_else
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                inputs:
                  - SignalCall:
                      signal:
                        id: on_off
                        scope: Local
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - on_off
                          - 0
                  - MapApplication:
                      function_expression:
                        Call:
                          id: " + "
                          typing:
                            Abstract:
                              - - Integer
                                - Integer
                              - Integer
                          location: ~
                      inputs:
                        - SignalCall:
                            signal:
                              id: counter
                              scope: Local
                            typing: Integer
                            location: ~
                            dependencies:
                              - - counter
                                - 0
                        - Constant:
                            constant:
                              Integer: 1
                            typing: Integer
                            location: ~
                            dependencies: []
                      typing: Integer
                      location: ~
                      dependencies:
                        - - counter
                          - 0
                  - Constant:
                      constant:
                        Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
                typing: Integer
                location: ~
                dependencies:
                  - - on_off
                    - 0
                  - - counter
                    - 0
            location: ~
        memory:
          buffers:
            memon_off:
              typing: Boolean
              initial_value:
                Boolean: true
              expression:
                SignalCall:
                  signal:
                    id: on_off
                    scope: Local
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - on_off
                      - 0
            memres:
              typing: Boolean
              initial_value:
                Boolean: true
              expression:
                MapApplication:
                  function_expression:
                    Call:
                      id: (_)
                      typing:
                        Abstract:
                          - - Boolean
                          - Boolean
                      location: ~
                  inputs:
                    - MapApplication:
                        function_expression:
                          Call:
                            id: " == "
                            typing:
                              Abstract:
                                - - Integer
                                  - Integer
                                - Boolean
                            location: ~
                        inputs:
                          - MapApplication:
                              function_expression:
                                Call:
                                  id: " + "
                                  typing:
                                    Abstract:
                                      - - Integer
                                        - Integer
                                      - Integer
                                  location: ~
                              inputs:
                                - SignalCall:
                                    signal:
                                      id: counter
                                      scope: Local
                                    typing: Integer
                                    location: ~
                                    dependencies:
                                      - - counter
                                        - 0
                                - Constant:
                                    constant:
                                      Integer: 1
                                    typing: Integer
                                    location: ~
                                    dependencies: []
                              typing: Integer
                              location: ~
                              dependencies:
                                - - counter
                                  - 0
                          - SignalCall:
                              signal:
                                id: tick_number
                                scope: Input
                              typing: Integer
                              location: ~
                              dependencies:
                                - - tick_number
                                  - 0
                        typing: Boolean
                        location: ~
                        dependencies:
                          - - counter
                            - 0
                          - - tick_number
                            - 0
                  typing: Boolean
                  location: ~
                  dependencies:
                    - - counter
                      - 0
                    - - tick_number
                      - 0
          called_nodes:
            counterocounter:
              node_id: counter
              signal_id: o
        location: ~
        graph:
          vertices:
            counter:
              id: counter
              value: White
              neighbors:
                - id: res
                  weight: 0
                - id: x
                  weight: 0
            memon_off:
              id: memon_off
              value: White
              neighbors: []
            memres:
              id: memres
              value: White
              neighbors: []
            on_off:
              id: on_off
              value: White
              neighbors:
                - id: res
                  weight: 0
                - id: memon_off
                  weight: 0
            res:
              id: res
              value: White
              neighbors:
                - id: memres
                  weight: 0
            status:
              id: status
              value: White
              neighbors:
                - id: on_off
                  weight: 0
                - id: counter
                  weight: 0
            tick_number:
              id: tick_number
              value: White
              neighbors: []
            x:
              id: x
              value: White
              neighbors: []
    location: ~
    graph:
      vertices:
        counter:
          id: counter
          value: Black
          neighbors:
            - id: res
              weight: 0
            - id: res
              weight: 1
        on_off:
          id: on_off
          value: Black
          neighbors:
            - id: res
              weight: 0
            - id: on_off
              weight: 1
        res:
          id: res
          value: Black
          neighbors:
            - id: counter
              weight: 1
            - id: tick_number
              weight: 1
        status:
          id: status
          value: Black
          neighbors:
            - id: on_off
              weight: 0
            - id: counter
              weight: 0
        tick_number:
          id: tick_number
          value: Black
          neighbors: []
component: ~
location: ~

